packages/block-editor/src/components/line-height-control/index.js

```diff
@@ -95,6 +95,50 @@ export default function LineHeightControl( {
 	__unstableHasLegacyStyles = true,
 } ) {
 	const isDefined = isLineHeightDefined( lineHeight );
+
+	const adjustNextValue = ( nextValue, wasTypedOrPasted ) => {
+		// Set the next value without modification if lineHeight has been defined
+		if ( isDefined ) return nextValue;
+
+		/**
+		 * The following logic handles the initial step up/down action
+		 * (from an undefined value state) so that the next values are better suited for
+		 * line-height rendering. For example, the first step up should immediately
+		 * go to 1.6, rather than the normally expected 0.1.
+		 *
+		 * Step up/down actions can be triggered by keydowns of the up/down arrow keys,
+		 * or by clicking the spin buttons.
+		 */
+		switch ( nextValue ) {
+			case `${ STEP }`:
+				// Increment by step value
+				return BASE_DEFAULT_VALUE + STEP;
+			case '0': {
+				// This means the user explicitly input '0', rather than stepped down
+				// from an undefined value state
+				if ( wasTypedOrPasted ) return nextValue;
+				// Decrement by step value
+				return BASE_DEFAULT_VALUE - STEP;
+			}
+			case '':
+				return BASE_DEFAULT_VALUE;
+			default:
+				return nextValue;
+		}
+	};
+
+	const stateReducer = ( state, action ) => {
+		// Be careful when changing this — cross-browser behavior of the
+		// `inputType` field in `input` events are inconsistent.
+		// For example, Firefox emits an input event with inputType="insertReplacementText"
+		// on spin button clicks, while other browsers do not even emit an input event.
+		const wasTypedOrPasted = [ 'insertText', 'insertFromPaste' ].includes(
+			action.payload.event.nativeEvent.inputType
+		);
+		state.value = adjustNextValue( state.value, wasTypedOrPasted );
+		return state;
```

Comment:
I'm not familiar with how exactly the `__unstableStateReducer` prop works, but is this is indeed a reducer, should we not create a new object instead of modifying the current state?

```suggestion
		return {
			...state,
			value: adjustNextValue( state.value, wasTypedOrPasted ),
		};
```

Reply:
I had the same exact thought, and was surprised to find that this implementation somehow requires the `state` object to remain stable. It actually does not work (!) if you return a fresh object. All the other instances in this reducer middleware chain does the same — mutates the same state object.

The reason was non-obvious to me, but I took another look today and found that `composeStateReducers()` is the [culprit](https://github.com/WordPress/gutenberg/blob/f9d97a71aeb0a4964cd6e17ed13ebb7f96f9650b/packages/components/src/input-control/reducer/reducer.ts#L54). The “reducer reducer” isn’t passing the accumulated state/action to each subsequent middleware function, but instead just passes references to the initial state/action objects. It’s unclear whether this design is intentional, but it’s definitely a departure from established conventions. I'll leave a TODO for potential reworking.

Reply:
>  It actually does not work (!) if you return a fresh object. All the other instances in this reducer middleware chain does the same — mutates the same state object.

🤯 

Thank you the sleuthing! It looks like this line [has been left untouched since the introduction of this component/code from the G2 repo](https://github.com/WordPress/gutenberg/pull/22329/files#diff-2162b9bdd5c3ff14bd49728fe9845ec80b3dca18fd0727bbabb9295ac7225bd2R65). We could do some further digging in that repo and see if there was any particular intention to implement the reducer composition this way.

Reply:
It kind of seems unintentional, but even if it were intentional I would prefer it changed because it's just so unexpected!

Reply:
Tracked here #38845